appserver/ejb/ejb-http-remoting/client/src/main/java/fish/payara/ejb/http/client/adapter/CompositeClientAdapter.java

```diff
@@ -0,0 +1,207 @@
+     * @throws NamingException if any of attempted client adapters throws one
+     */
+    @Override
+    public Optional<Object> makeLocalProxy(String jndiName, Context remoteContext) throws NamingException {
+        Optional<ResolutionResult> resolutionResult = adapterSuppliers.stream()
+                .map(Supplier::get)
+                .map(adapter -> resolve(adapter, jndiName, remoteContext))
+                .filter(Optional::isPresent)
+                .findFirst()
+                .map(Optional::get);
+        if (resolutionResult.isPresent()) {
+            ResolutionResult result = resolutionResult.get();
+            if (result.exception != null) {
+                throw result.exception;
+            } else {
+                return Optional.of(result.instance);
+            }
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Convenience method for constructing a {@link Builder}
+     * @return new builder instance
+     */
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    private Optional<ResolutionResult> resolve(ClientAdapter adapter, String jndiName, Context remoteContext) {
+        try {
+            return adapter.makeLocalProxy(jndiName, remoteContext).map(ResolutionResult::result);
+        } catch (NamingException e) {
+            return Optional.of(ResolutionResult.exception(e));
+        }
+    }
+
+    private static class ResolutionResult {
+        Object instance;
+        NamingException exception;
+
+        static ResolutionResult result(Object instance) {
+            ResolutionResult result = new ResolutionResult();
+            result.instance = instance;
+            return result;
+        }
+
+        static ResolutionResult exception(NamingException exception) {
+            ResolutionResult result = new ResolutionResult();
+            result.exception = exception;
+            return result;
+        }
+    }
+
+
+    /**
+     * Builder for composite Client Adapter.
+     * <p>Adapters can be registered in three ways:</p>
+     * <ul>
+     *     <li>{@link #register(ClientAdapter...)} : providing an instance</li>
+     *     <li>{@link #register(Class[])} : providing a class name. Class will be immediately instantiated via default constructor. </li>
+     *     <li>{@link #register(Supplier[])} : providing instance supplier. The supplier will be invoked for each lookup. This can enable more contextual
+     *          use cases, for example thread-local adapters.</li>
+     * </ul>
+     * Further customization of client adapter is possible by means of {@link ClientAdapterCustomizer}
+     */
+    public static final class Builder {
+        private List<Supplier<? extends ClientAdapter>> adapterSuppliers = new ArrayList<>();
+
+        /**
+         * Build resulting adapter.
+         * @return new instance of CompositeClientAdapter
+         */
+        public CompositeClientAdapter build() {
```

Comment:
For consideration: Since the end result of the `build`ing is a `CompositeClientAdapter` the `register` methods could equally be directly on the `CompositeClientAdapter` making the `Builder` class unnecessary.

Reply:
I was aiming for immutability of `CompositeClientAdapter`. Allowing the list of downstream suppliers to change at runtime would add even more dynamics into already dynamic matching process. 

Immutability could be also achieved with `register` methods returning new instances of `CompositeClientAdapter`, but it would be confusing. Naming the method `withAdapters` or similar may clear that confusion, but Builders are quite common in newer Jakarta APIs, so I think there's less element of surprise in this setup.